get rid of gms/failure_detector

bhalevy opened the pull request:
Move gms::arrival_window to api/failure_detector which is its only user.
and get rid of the rest, which is not used, now that we use direct_failure_detector instead.

TODO: integare direct_failure_detector with failure_detector api.

tgrabiec commented on test/rest_api/test_failure_detector.py:
```diff
@@ -0,0 +1,18 @@
+    resp.raise_for_status()
+    assert len(resp.json()) == 1
+    addr = resp.json()[0]['addrs']
+    re.match(r'\d+\.\d+\.\d+\.\d+', addr)
```
Result of match() is ignored, so the statement has no effect.

bhalevy replied:
fixed in v2

tgrabiec commented on api/failure_detector.cc:
```diff
@@ -8,9 +8,94 @@
+
+namespace gms {
+
+class arrival_window {
```
The only use is here:

```

    fd::get_endpoint_phi_values.set(r, [](std::unique_ptr<request> req) {
        std::map<gms::inet_address, gms::arrival_window> map;
        std::vector<fd::endpoint_phi_value> res;
        auto now = gms::arrival_window::clk::now();
        for (auto& p : map) {
            fd::endpoint_phi_value val;
            val.endpoint = p.first.to_sstring();
            val.phi = p.second.phi(now);
            res.emplace_back(std::move(val));
        }
        return make_ready_future<json::json_return_type>(res);
    });
```

Notice that `map` is not initialized, so it's empty. Before 0ef33b71bab2b15f8f42acea44fd4049fafe46e2, it was obtained from the gossiper gms::get_arrival_samples().

So you can just remove arrival_window altogether.

I don't think the logic of arrival_window belongs to api, it belongs to the failure detector. If there is no longer a failure detector, there should be no arrival_window.

bhalevy replied:
fixed in v2

bhalevy commented:
In v2 ([cfe69f5](https://github.com/scylladb/scylladb/commit/cfe69f59e52beaa9b3200ee7cc0a48ea13950099)):
- fixed unit test to assert the value of re.match
- removed degenerate use of failure_dector::arrival_window in api
- got rid of gms/failure_detector.{cc,hh} as a whole

kbr-scylla commented on api/failure_detector.cc:
```diff
@@ -81,15 +80,10 @@ void set_failure_detector(http_context& ctx, routes& r, gms::gossiper& g) {
-        std::map<gms::inet_address, gms::arrival_window> map;
+        // FIXME: this was based on gms::get_arrival_samples.
+        // But the latter returned an empty std::map<inet_address, arrival_window>.
+        // The functionality needs to be implemented by the failure detector.
```
We no longer have a phi failure detector, so the functionality will never be implemented.
Maybe we should just remove the API. Even if someone depends on it (I doubt it) it wasn't returning anything meaningful for some time already.

asias replied:
We do not need the FIXME, just returning the empty value is good enough.  The phi values associated with the old fd is gone and is not going back.

avikivity commented:
@asias please review too

asias commented:
Looks good. I have a minor comment inline.

bhalevy commented:
In v3 ([3f1ac84](https://github.com/scylladb/scylladb/commit/3f1ac846d85fd1df12f0e631e11d504cc29e0a7a)):
- rebased
- turned FIXME comment into a regular one
  - just explaining why the api function returns an empty value

scylladb-promoter commented:
CI state **`SUCCESS`** - https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/785/

kbr-scylla commented:
Qd
